Kyle0654 inpaint improvement - with refactoring from PR #1221

lstein opened the pull request:
# I made a few suggested modifications to your PR:
    
1. pass a PIL.Image to img2img and inpaint rather than tensor. This avoids passing two extra arguments and cleans up generate() (a bit)
2. Clipseg generates an "L" or "1" format mask, so the mask won't necessarily have any transparency. Instead of getting the mask's "A" channel, just convert it to "L".
3. Moved the postprocessing code into an overridden sample_to_image() method, keeping the generation code separate from the image processing code.

BTW, the CLI tries to autogenerate masks from images that contain transparency so that you don't have to pass both the original and the transparent version. When using this with your unmodified PR, I got a crash:

```
 ╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /u/lstein/projects/SD/InvokeAI/./scripts/invoke.py:855 in <module>                               │
│                                                                                                  │
│   852 ######################################                                                     │
│   853                                                                                            │
│   854 if __name__ == '__main__':                                                                 │
│ ❱ 855 │   main()                                                                                 │
│   856                                                                                            │
│                                                                                                  │
│ /u/lstein/projects/SD/InvokeAI/./scripts/invoke.py:104 in main                                   │
│                                                                                                  │
│   101 │   │   )                                                                                  │
│   102 │                                                                                          │
│   103 │   try:                                                                                   │
│ ❱ 104 │   │   main_loop(gen, opt)                                                                │
│   105 │   except KeyboardInterrupt:                                                              │
│   106 │   │   print("\ngoodbye!")                                                                │
│   107                                                                                            │
│                                                                                                  │
│ /u/lstein/projects/SD/InvokeAI/./scripts/invoke.py:330 in main_loop                              │
│                                                                                                  │
│   327 │   │   │   if operation == 'generate':                                                    │
│   328 │   │   │   │   catch_ctrl_c = infile is None # if running interactively, we catch keybo   │
│   329 │   │   │   │   opt.last_operation='generate'                                              │
│ ❱ 330 │   │   │   │   gen.prompt2image(                                                          │
│   331 │   │   │   │   │   image_callback=image_writer,                                           │
│   332 │   │   │   │   │   step_callback=step_callback,                                           │
│   333 │   │   │   │   │   catch_interrupts=catch_ctrl_c,                                         │
│                                                                                                  │
│ /u/lstein/projects/SD/InvokeAI/ldm/generate.py:396 in prompt2image                               │
│                                                                                                  │
│    393 │   │   │   │   log_tokens    =self.log_tokenization                                      │
│    394 │   │   │   )                                                                             │
│    395 │   │   │                                                                                 │
│ ❱  396 │   │   │   init_image,mask_image,pil_image,pil_mask = self._make_images(                 │
│    397 │   │   │   │   init_img,                                                                 │
│    398 │   │   │   │   init_mask,                                                                │
│    399 │   │   │   │   width,                                                                    │
│                                                                                                  │
│ /u/lstein/projects/SD/InvokeAI/ldm/generate.py:655 in _make_images                               │
│                                                                                                  │
│    652 │   │   elif text_mask:                                                                   │
│    653 │   │   │   init_mask = self._txt2mask(image, text_mask, width, height, fit=fit)          │
│    654 │   │                                                                                     │
│ ❱  655 │   │   return init_image, init_mask, image, mask_image                                   │
│    656 │                                                                                         │
│    657 │   def _make_base(self):                                                                 │
│    658 │   │   if not self.generators.get('base'):                                               │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
```

wfng92 committed:
Removed duplicate fix_func for MPS

lstein committed:
add support for loading VAE autoencoders

To add a VAE autoencoder to an existing model:

1. Download the appropriate autoencoder and put it into
   models/ldm/stable-diffusion

   Note that you MUST use a VAE that was written for the
   original CompViz Stable Diffusion codebase. For v1.4,
   that would be the file named vae-ft-mse-840000-ema-pruned.ckpt
   that you can download from https://huggingface.co/stabilityai/sd-vae-ft-mse-original

2. Edit config/models.yaml to contain the following stanza, modifying `weights`
   and `vae` as required to match the weights and vae model file names. There is
   no requirement to rename the VAE file.

~~~
stable-diffusion-1.4:
  weights: models/ldm/stable-diffusion-v1/sd-v1-4.ckpt
  description: Stable Diffusion v1.4
  config: configs/stable-diffusion/v1-inference.yaml
  vae: models/ldm/stable-diffusion-v1/vae-ft-mse-840000-ema-pruned.ckpt
  width: 512
  height: 512
~~~

3. Alternatively from within the `invoke.py` CLI, you may use the command
   `!editmodel stable-diffusion-1.4` to bring up a simple editor that will
   allow you to add the path to the VAE.

4. If you are just installing InvokeAI for the first time, you can also
   use `!import_model models/ldm/stable-diffusion/sd-v1.4.ckpt` instead
   to create the configuration from scratch.

5. That's it!

lstein committed:
Merge branch 'development' into patch-1

lstein committed:
Merge branch 'inpaint-improvement' of https://github.com/Kyle0654/InvokeAI into Kyle0654-inpaint-improvement

lstein committed:
ported code refactor changes from PR #1221

- pass a PIL.Image to img2img and inpaint rather than tensor
- To support clipseg, inpaint needs to accept an "L" or "1" format
  mask. Made the appropriate change.

lstein commented:
I'm afraid some other stuff got in there because of divergence between your fork and mine. The only files that changed are:
1. ldm/generate.py
2. ldm/invoke/generator/img2img.py
3. ldm/invoke/generator/inpaint.py

Kyle0654 commented on ldm/invoke/generator/inpaint.py:
```diff
@@ -96,30 +104,46 @@ def make_image(x_T):
-            # Get numpy version
-            np_gen_result = np.asarray(gen_result, dtype=np.uint8)
+        # Get the original alpha channel of the mask
+        pil_init_mask = pil_mask.convert('L')
```
The UI is currently passing RGBA as the mask. Could call it a bug I guess, and make the UI fix it.

Kyle0654 commented on ldm/invoke/generator/inpaint.py:
```diff
@@ -31,17 +35,21 @@ def get_make_image(self,prompt,sampler,steps,cfg_scale,ddim_eta,
-        init_a_pixels = np.asarray(pil_init_image.getchannel('A'), dtype=np.uint8).reshape(pil_init_mask.width * pil_init_mask.height)
-        init_rgb_pixels = init_rgb_pixels[init_a_pixels > 0]
-        init_rgb_pixels = init_rgb_pixels.reshape(1, init_rgb_pixels.shape[0], init_rgb_pixels.shape[1]) # Filter to just pixels that have any alpha, this is now our histogram
+        if isinstance(init_image, PIL.Image.Image):
```
If `init_image` isn't a Pillow image, we won't set `pil_image`, and `sample_to_image` below will break. Should we either assert it's a Pillow image here, or assume it is (possibly adding a type hint to the function definition) and let it fail if not?

Kyle0654 replied:
Same issue with `mask_image` right below this

lstein replied:
Good point. In my version this case would just fall back to the previous behaviour and not try the post processing. I’ll adjust to do the same.

lstein replied:
I've fixed. In the event that a tensor (or something else) is passed, the post-generation fixup code will not execute and behavior will be as before.

Kyle0654 commented on ldm/invoke/generator/inpaint.py:
```diff
@@ -96,30 +104,46 @@ def make_image(x_T):
-            # Get numpy version
-            np_gen_result = np.asarray(gen_result, dtype=np.uint8)
+        # Get the original alpha channel of the mask
+        pil_init_mask = pil_mask.convert('L')
```
The UI is currently passing RGBA as the mask. Could call it a bug I guess, and make the UI fix it.

lstein replied:
How about we check and convert as necessary?

lstein replied:
Looks for RGB/L vs RGBA and responds appropriately.

lstein committed:
minor fixes to inpaint code

1. If tensors are passed to inpaint as init_image and/or init_mask, then
   the post-generation image fixup code will be skipped.

2. Post-generation image fixup will work with either a black and white "L"
   or "RGB"  mask, or an "RGBA" mask.
